### ISSUE TITLE
Introduce a convert method that does not rely on an Asciidoctor::Document object

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor-diagram#206]
I think we should introduce a `convert` method with the following signature:

```rb
# Convert the textual representation of a diagram to an image as a String.
#
# input   - the String textual representation of the diagram
# opts    - an Hash of options to control processing (e.g., output format (:format), custom attributes (:attributes)).
#
# Returns the diagram image as a String
def convert source, opts
```

Arguably the `format` could be an argument.
Once we have integrated this method into the code, it should be easier to add a server: https://github.com/asciidoctor/asciidoctor-diagram/issues/183

**Comment by pepijnve** [comment]
@Mogztter sorry for the delay in response. I haven't been able to keep up with my projects during the christmas holidays.

Thanks for the offer to help out with this, that's greatly appreciated. The main refactoring I had in mind myself was to try to decouple the image conversion logic itself from the Asciidoctor API so that it could be reused in a service context as well. I had some vague ideas about this change, but nothing fully formed yet. I suggest we start from your proposal above and flesh that out.

The way the code is currently structured each extension provides one or more block or block macro processor classes. Each of these typically classes then includes a module that provides the actual conversion logic. When the conversion logic modules are included they register Procs with the base DiagramProcessor class key'ed by image output format. When the block/macro processing code is invoked a lookup based on image format is performed and then the registered Proc (if any) gets invoked. All of this is just boilerplate code to set up the the Asciidoctor extensions with minimal repeated logic.

What I think you're suggesting is that instead of having those ad-hoc Procs, a uniform interface is introduced for all the image generation modules instead. Sounds like a good first step indeed.

I would start by splitting the boilerplate code I mentioned above and image generation logic itself into two distinct parts. Using PlantUML as an example, split off the [plantuml method](https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/plantuml/extension.rb#L23) into its own module or class in a separate file.

The second step is to then adapt the plantuml method to match your proposal. I think you'll run into some issues here. Most of the image conversion logic makes use of some facilities that are provided by Asciidoctor (path resolution, looking up attributes in the node tree, etc). A simple Hash of options is probably not sufficiently rich for this.
At the moment [DiagramSource](https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/extensions.rb#L336) provides some, but not all, of the needed functionality. The source object that's passed to the Procs is actually an instance of DiagramSource, not a String. So perhaps the easiest way forward is to extend DiagramSource with the missing bits of functionality that the image generators need.
The convert signature would then become

```
# Convert the textual representation of a diagram to an image as a String.
#
# input   - a DiagramSource, the source code representation of the diagram
# opts    - a Hash of options to control processing (e.g., output format (:format)).
#
# Returns the diagram image as a String
def convert source, opts
```

Does that make sense?

**Comment by Mogztter** [comment]
> sorry for the delay in response. I haven't been able to keep up with my projects during the christmas holidays.

No worries, happy new year by the way 🎉 

> Thanks for the offer to help out with this, that's greatly appreciated. The main refactoring I had in mind myself was to try to decouple the image conversion logic itself from the Asciidoctor API so that it could be reused in a service context as well. I had some vague ideas about this change, but nothing fully formed yet. I suggest we start from your proposal above and flesh that out.

That's what I had in mind.
I've also been thinking a lot about the "diagram server". At first I thought about using a Java with Jython and/or GraalVM to create a Polyglot application. I thought it would be easy to integrate with BlockDiag (Python) and Mermaid (JavaScript) but that's not the case. BlockDiag is using system library (zlib, jpeglib, tiff...) and Mermaid needs a DOM 😞 

Anyway let's focus on this decoupling first 💪 

> What I think you're suggesting is that instead of having those ad-hoc Procs, a uniform interface is introduced for all the image generation modules instead. Sounds like a good first step indeed.

Yes 👍 
Once we have this common interface, it's easy to delegate to a third-party server.

> I would start by splitting the boilerplate code I mentioned above and image generation logic itself into two distinct parts. Using PlantUML as an example, split off the plantuml method into its own module or class in a separate file.

Indeed, we should split one diagram processor at a time otherwise it will be really hard to review and potential regressions will be harder to fix.

> The second step is to then adapt the plantuml method to match your proposal. I think you'll run into some issues here. Most of the image conversion logic makes use of some facilities that are provided by Asciidoctor (path resolution, looking up attributes in the node tree, etc). A simple Hash of options is probably not sufficiently rich for this.

I've already run into some issues 😀 

> At the moment DiagramSource provides some, but not all, of the needed functionality. The source object that's passed to the Procs is actually an instance of DiagramSource, not a String. So perhaps the easiest way forward is to extend DiagramSource with the missing bits of functionality that the image generators need.
The convert signature would then become

I think at the end, we will still need to use simple objects (String, Map, Hash) for serialization.
My end goal is to provide an HTTP service (hosted or self-hosted) with a standard API 💥 

I think we agree on this decoupling/refactoring. I'll start with the Plantuml method and open a pull request, stay tuned 😉

**Comment by mojavelinux** [comment]
> Pepijn wrote: The main refactoring I had in mind myself was to try to decouple the image conversion logic itself from the Asciidoctor API so that it could be reused in a service context as well.

:+1: 

> Guillaume wrote: I've also been thinking a lot about the "diagram server".

The impact of such a tool would be far reaching. It could also help get more people involved as its use wouldn't be limited to just AsciiDoc / Asciidoctor.

> def convert source, opts

I would say make sure to leave room for passing in information about the context (the current document, for example). I've seen APIs that try to focus exclusively on the source (Jekyll comes to mind) and then you end up not being able to handle things like references within the source. Perhaps it's an option. Or perhaps a context argument?

**Comment by pepijnve** [comment]
FYI, this is being worked on in https://github.com/asciidoctor/asciidoctor-diagram/tree/server_prep
The converters have already been put in place. Still need to tidy up a bunch of code and implement the server frontend.

**Comment by pepijnve** [comment]
The code for all this is integrated on `master` and will be included in v2.0.0
